modules/apps/transfer/types/packet.go

```diff
@@ -64,3 +68,119 @@ func (ftpd FungibleTokenPacketData) ValidateBasic() error {
+to retrieve the desired callback addresses for the ICS20 packet on the source and destination chains.
+
+The Memo is used to ensure that the callback is desired by the user. This allows a user to send an ICS20 packet
+to a contract with ADR-8 enabled without automatically triggering the callback logic which may lead to unexpected
+behaviour.
+
+The Memo format is defined like so:
+
+```json
+{
+	// ... other memo fields we don't care about
+	"callback": {
+		"src_callback_address": {contractAddrOnSourceChain},
+		"dest_callback_address": {contractAddrOnDestChain},
+
+		// optional fields
+		"gas_limit": {intForCallback}
+	}
+}
+```
+
+For transfer, we will NOT enforce that the src_callback_address is the same as sender and dest_callback_address is the same as receiver.
+
+*/
+
+// GetSourceCallbackAddress returns the callback address if it is specified in
+// the packet data memo. If no callback address is specified, an empty string is returned.
+//
+// The memo is expected to contain the source callback address in the following format:
+// { "callback": { "src_callback_address": {contractAddrOnSourceChain}}
+//
+// ADR-8 middleware should callback on the returned address if it is a PacketActor
+// (i.e. smart contract that accepts IBC callbacks).
+func (ftpd FungibleTokenPacketData) GetSourceCallbackAddress() string {
+	callbackData := ftpd.getCallbackData()
+	if callbackData == nil {
+		return ""
+	}
+
+	srcCallbackAddress, ok := callbackData["src_callback_address"].(string)
+	if !ok {
+		return ""
+	}
+
+	return srcCallbackAddress
+}
+
+// GetDestCallbackAddress returns the callback address if it is specified in
+// the packet data memo. If no callback address is specified, an empty string is returned.
+//
+// The memo is expected to contain the destination callback address in the following format:
+// { "callback": { "dest_callback_address": {contractAddrOnDestChain}}
+//
+// ADR-8 middleware should callback on the returned address if it is a PacketActor
+// (i.e. smart contract that accepts IBC callbacks).
+func (ftpd FungibleTokenPacketData) GetDestCallbackAddress() string {
+	callbackData := ftpd.getCallbackData()
+	if callbackData == nil {
+		return ""
+	}
+
+	destCallbackAddress, ok := callbackData["dest_callback_address"].(string)
+	if !ok {
+		return ""
+	}
+
+	return destCallbackAddress
+}
+
+// UserDefinedGasLimit returns the custom gas limit provided in the packet data memo.
+//
+// The memo is expected to specify the callback address in the following format:
+// { "callback": { ... , "gas_limit": {intForCallback} }
+//
+// If no gas limit is specified, 0 is returned.
+func (ftpd FungibleTokenPacketData) UserDefinedGasLimit() uint64 {
+	callbackData := ftpd.getCallbackData()
+	if callbackData == nil {
+		return 0
+	}
+
+	// json number won't be unmarshaled as a uint64, so we need to cast it to float64 first
+	gasLimit, ok := callbackData["gas_limit"].(float64)
```

Comment:
we can require specifying the number as a string and unmarshalling it as a decimal. I think this should be safer as floats can behave differently on different architectures

Reply:
This is a good idea. Will implement.